book/src/proposals/leader-duplicate-block-slashing.md

```diff
@@ -0,0 +1,219 @@
+
+* signature
+* header: slot index, shred index
+* message data
+
+The signature is of the merkle tree of the shred data.
+
+```text
+              merke root
+    /                        \
+(slot index, shred index)        hash(msg)
+```
+
+## Proof of a Duplicate Shred
+
+Any two different signatures that show a merkle path to the same
+`(slot index, shred index)` for the same leader are proof that the
+leader produced two conflicting blocks.
+
+## Storing Multiple Versions of the Same Slot
+When there is a possibility of duplicate blocks, a validator need to track and
+replay multiple versions of a slot in case any one of these versions is 
+finalized by the cluster.
+
+### Generating a Unique Identifier for Each Version of a Slot
+We distinguish different versions of a slot by the ending blockhash. 
+This blockhash thus needs to be "functionally unique" (meaning a malicious
+leader cannot grind out a duplicate in the amount of time they have to
+generate a block) across all possible shred outputs generated by a leader. 
+
+To this end it's important to note that:
+1) Two different versions of a slot must generate a different set of shreds.
+Two different versions of a slot must either:
+    a) Chain from different parents, in which case the PoH of the entries will 
+    different
+    b) Contain different set of transactions, in which case the PoH of the 
+    entries will be different
+
+2) The resulting merkle trees of two different sets of shreds have 
+"functionally unique" merkle roots.
+
+Then if we were to generate a merkle tree out of the shred batches for a slot, 
+and define the blockhash of a slot to be:
+
+ `hash(last tick, shred merkle root)`
+
+This blockhash should be "functionally unique".
+
+### Generating the Merkle Tree for the Shreds
+1) Leader blocks on Poh before last tick can be generated
+2) Leader waits for BroadcastStage to finish generating all the shreds
+3) Computes the Merkle tree of all the shreds
+4) Inserts an entry into the Poh Stream with the Merkle root
+5) Restart PoH
+
+### Indexing the Column Families by Blockhash
+We augment all slot-related column families in blocktree to include a blockhash
+to support tracking multiple versions of the same slot. We outline the notable
+ones below:
+
+1) `Data` and `Erasure` column families: `(slot, blockhash, shred_index)`
+Two cases here:
+   a) When leaders generate their own slot, the `blockhash` portion of the key
+   is set to `Hash::default()` until the blockhash is computed at the end of 
+   the slot. Then the leader will have to store a mapping from 
+   `Hash::default()` to the actual blockhash in a separate area of storage 
+   (another column family?). This is important in order to respond to repairs
+   which will now specify a `blockhash` in addtion to a `slot` and `index`. 
+   Thus if the node crashes before this mapping can be stored, then on restart
+   the validator needs to recompute the blockhash in `blocktree_processor`.
```

Comment:
One downside of this approach is that if we have two versions of a slot, each with 100 shreds, the first 99 of which are the same, and the last different, we have to store 200 copies.